keps/sig-storage/2644-honor-pv-reclaim-policy/README.md

```diff
@@ -0,0 +1,396 @@
+	klog.V(4).Infof("deleteVolumeOperation [%s] started", volume.Name)
+	var requestToDeleteAnn bool
+	// This method may have been waiting for a volume lock for some time.
+	// Previous deleteVolumeOperation might just have saved an updated version, so
+	// read current volume state now.
+	newVolume, err := ctrl.kubeClient.CoreV1().PersistentVolumes().Get(context.TODO(), volume.Name, metav1.GetOptions{})
+	if err != nil { //<===============================Changed here START
+		if !apierrors.IsNotFound(err) {
+			return "", nil
+		} else {
+			// Check if the volume is annotated with explicit request to delete.
+			// In this case due to race with the pv protection controller, it was not possible to retrieve the latest PV
+			// state. However, the annotation clearly indicates the volume deletion can proceed.
+			requestToDeleteAnn = metav1.HasAnnotation(volume.ObjectMeta, pvutil.AnnRequestToDeleteVolume)
+			if !requestToDeleteAnn {
+				klog.Infof("no request to delete annotation found on the volume %s", volume.Name)
+				// Does not have the annotation, throw the NotFound error.
+				return "", err
+			} else {
+				klog.Infof("request to delete annotation found on the volume %s", volume.Name)
+				newVolume = volume
+			}
+		}
+	} //<===============================Changed here END
+    // Note: no need to check for deletion timestamp, the schedule operation ensures that duplicate delete
+    // are not processed.
+    if newVolume.GetDeletionTimestamp() != nil {//<===========================Check to be removed
+        klog.V(3).Infof("Volume %q is already being deleted", volume.Name)
+        return "", nil
+    }
+// Remaining code below skipped for brevity...
+// Please refer pkg/controller/volume/persistentvolume/pv_controller.go in kubernetes/kubernetes for the full code.
+```
+
+The above changes ensure that the plugin receives the volume deletion request.
+
+On the driver side, additional checks are made whenever `DeletionTimestamp` is set.
+
+```go
+// shouldDelete returns whether a volume should have its backing volume
+// deleted, i.e. whether a Delete is "desired"
+func (ctrl *ProvisionController) shouldDelete(ctx context.Context, volume *v1.PersistentVolume) bool {
+	if deletionGuard, ok := ctrl.provisioner.(DeletionGuard); ok {
+		if !deletionGuard.ShouldDelete(ctx, volume) {
+			return false
+		}
+	}
+
+	// In 1.9+ PV protection means the object will exist briefly with a
+	// deletion timestamp even after our successful Delete. Ignore it.
+	if ctrl.kubeVersion.AtLeast(utilversion.MustParseSemantic("v1.9.0")) {
+		if ctrl.addFinalizer && !ctrl.checkFinalizer(volume, finalizerPV) && volume.ObjectMeta.DeletionTimestamp != nil {
+			return false
+		} else if volume.ObjectMeta.DeletionTimestamp != nil {
+			// Check if there is a request to delete the volume explicitly.
+			if !metav1.HasAnnotation(volume.ObjectMeta, annRequestToDeleteVolume) { //<===============================Changed here START
```

Comment:
Again, API server may have already deleted the PV here, as all finalizers may have been already removed and the external-provisioner may have been restarted or slow to process events. I am afraid the solution described here is not robust enough and you will need to introduce a new finalizer to make 100% sure that a PV cannot be deleted from the API server before the external provisioner confirms it has deleted the volume.

Reply:
Thanks for taking a look, I've updated the KEP with the finalizer approach.